### DISCUSSION TITLE
eglot codeAction could add a `context` key

### DISCUSSION THREAD

**Discussion by Andre0991** [joaotavora/eglot#895] [Ideas]
Hi.

It seems that `eglot` does not send `context` in [`codeAction`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction) requests. This prevent the server from responding with codeActions that are available for the same line when the point is not necessarily over some diagnostics.

According to the spec,

```
export interface CodeActionContext {
	/**
	 * An array of diagnostics known on the client side overlapping the range
	 * provided to the `textDocument/codeAction` request. They are provided so
	 * that the server knows which errors are currently presented to the user
	 * for the given range. There is no guarantee that these accurately reflect
	 * the error state of the resource. The primary parameter
	 * to compute code actions is the provided range.
	 */
```

For example, consider this sample `clojure` file:

```clojure
(ns foo)

(str/foo) bar
```

Suppose the point is in the last position of the last line, just after `bar`.

Then, `eglot` sends this `codeAction` request:

```
(:jsonrpc "2.0" :id 11 :method "textDocument/codeAction" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj")
	   :range
	   (:start
	    (:line 2 :character 12)
	    :end
	    (:line 2 :character 12))
	   :context
	   (:diagnostics
	    [])))
```

and this is the server reply, containing only one action:

```
(:jsonrpc "2.0" :id 11 :result
	  [(:title "Clean namespace" :kind "source.organizeImports" :command
		   (:title "Clean namespace" :command "clean-ns" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 12]))])
```

Now, if `eglot` sent context, it would return more code actions.

For example, this is a sample from `lsp-mode`, with the same file:

```
{
  "textDocument": {
    "uri": "file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj"
  },
  "range": {
    "start": {
      "line": 2,
      "character": 12
    },
    "end": {
      "line": 2,
      "character": 12
    }
  },
  "context": {
    "diagnostics": [
      {
        "range": {
          "start": {
            "line": 2,
            "character": 1
          },
          "end": {
            "line": 2,
            "character": 8
          }
        },
        "severity": 2,
        "code": "unresolved-namespace",
        "source": "clj-kondo",
        "message": "Unresolved namespace str. Are you missing a require?",
        "tags": []
      }
    ]
  }
}
```

and the result contains many actions (I'll omit it here because it's too big, it's suggesting lots of imports whose name could be `str`).

---

I think the most annoying situation that I miss without context is not being able to use an import code action when the point is just after part of a symbol.

For example, say the line is

```
(str/starts-with|)
```

where `|` is the point. It would like to use `eglot-code-actions` for importing `str` as `[clojure.string :as str]` just after pressing h. However, I need to move backward one char, moving the point above the missing import diagnostic. Only then eglot displays the code action that I need.

**Comment by joaotavora** [comment]
As far as I can read from the code, yes Eglot sends context, of course it does. 

```lisp
(jsonrpc-request
           server
           :textDocument/codeAction
           (list :textDocument (eglot--TextDocumentIdentifier)
                 :range (list :start (eglot--pos-to-lsp-position beg)
                              :end (eglot--pos-to-lsp-position end))
                 :context
                 `(:diagnostics
                   [,@(cl-loop for diag in (flymake-diagnostics beg end)
                               when (cdr (assoc 'eglot-lsp-diag
                                                (eglot--diag-data diag)))
                               collect it)]
                   ,@(when action-kind `(:only [,action-kind]))))
           :deferred t)
```


 The "context" that the `lsp-mode` is giving does contain a diagnostic.  Is at least that diagnostic in the Eglot version as well? There's no way to tell from the information you gave us.

Anyway, if it is then it is possible to transmit it to the server, but that depends how you invoke `eglot-code-actions`.  How did you do it?  Did you `M-x eglot-code-actions` with a region selected or just a specific point?  

Reading your log, it seems the latter.  It seems that `lsp-mode` liberally expands the context collection range to the span at least the whole line after being asked to provide code actions on a point.  We could do that (assuming it's LSP-kosher) but I would like to know what happens in your example if you select the whole line and _then_ `M-x eglot-code-actions`.

**Reply by Andre0991** [reply]
> Is at least that diagnostic in the Eglot version as well? There's no way to tell from the information you gave us.

No, `eglot` is sending empty diagnostics as context – this is present in the logs I put in the discussion.

```
:context
	   (:diagnostics
	    [])
```

> Did you M-x eglot-code-actions with a region selected or just a specific point?

Just a specific point.

> We could do that (assuming it's LSP-kosher) but I would like to know what happens in your example if you select the whole line and then M-x eglot-code-actions.

I just tested selecting the whole line – I get all suggestions. Logs:

<details>

```
[client-request] (id:54) Thu Mar 24 09:20:22 2022:
(:jsonrpc "2.0" :id 54 :method "textDocument/codeAction" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj")
	   :range
	   (:start
	    (:line 2 :character 0)
	    :end
	    (:line 2 :character 13))
	   :context
	   (:diagnostics
	    [(:range
	      (:start
	       (:line 2 :character 1)
	       :end
	       (:line 2 :character 8))
	      :severity 2 :code "unresolved-namespace" :source "clj-kondo" :message "Unresolved namespace str. Are you missing a require?" :tags
	      [])])))
[server-reply] (id:54) Thu Mar 24 09:20:22 2022:
(:jsonrpc "2.0" :id 54 :result
	  [(:title "Add require '[clojure.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add missing require" :command "add-missing-libspec" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1]))
	   (:title "Add require '[common-core.streams :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "common-core.streams" "str" nil]))
	   (:title "Add require '[cider.nrepl.inlined-deps.rewrite-cljs.v0v4v4.rewrite-clj.node.stringz :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "cider.nrepl.inlined-deps.rewrite-cljs.v0v4v4.rewrite-clj.node.stringz" "str" nil]))
	   (:title "Add require '[common-core.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "common-core.string" "str" nil]))
	   (:title "Add require '[manifold.stream :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "manifold.stream" "str" nil]))
	   (:title "Add require '[clojure.core.strint :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "clojure.core.strint" "str" nil]))
	   (:title "Add require '[cider.nrepl.inlined-deps.rewrite-clj.v0v6v1.rewrite-clj.node.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "cider.nrepl.inlined-deps.rewrite-clj.v0v6v1.rewrite-clj.node.string" "str" nil]))
	   (:title "Add require '[clojure.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "clojure.string" "str" nil]))
	   (:title "Add require '[camel-snake-kebab.internals.string-separator :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "camel-snake-kebab.internals.string-separator" "str" nil]))
	   (:title "Add require '[cider.nrepl.inlined-deps.rewrite-cljs.v0v4v4.rewrite-clj.parser.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "cider.nrepl.inlined-deps.rewrite-cljs.v0v4v4.rewrite-clj.parser.string" "str" nil]))
	   (:title "Add require '[cider.nrepl.inlined-deps.rewrite-clj.v0v6v1.rewrite-clj.parser.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "cider.nrepl.inlined-deps.rewrite-clj.v0v6v1.rewrite-clj.parser.string" "str" nil]))
	   (:title "Add require '[flare.string :as str]'" :kind "quickfix" :isPreferred t :command
		   (:title "Add require suggestion" :command "add-require-suggestion" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "flare.string" "str" nil]))
	   (:title "Create namespace 'str' and 'foo' function" :kind "quickfix" :command
		   (:title "Create function" :command "create-function" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1]))
	   (:title "Change coll to vector" :kind "refactor" :command
		   (:title "Change coll" :command "change-coll" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0 "vector"]))
	   (:title "Change coll to set" :kind "refactor" :command
		   (:title "Change coll" :command "change-coll" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0 "set"]))
	   (:title "Change coll to map" :kind "refactor" :command
		   (:title "Change coll" :command "change-coll" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0 "map"]))
	   (:title "Thread first all" :kind "refactor.rewrite" :command
		   (:title "Thread first all" :command "thread-first-all" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0]))
	   (:title "Thread last all" :kind "refactor.rewrite" :command
		   (:title "Thread last all" :command "thread-last-all" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0]))
	   (:title "Suppress 'unresolved-namespace' diagnostic" :kind "quickfix" :command
		   (:title "Suppress diagnostic" :command "suppress-diagnostic" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 1 "unresolved-namespace"]))
	   (:title "Clean namespace" :kind "source.organizeImports" :command
		   (:title "Clean namespace" :command "clean-ns" :arguments
			   ["file:///Users/andreperictavares/dev/nu/isa-bff/src/isa_bff/controllers/service/foo.clj" 2 0]))])
```

</details>

**Reply by Andre0991** [reply]
>  It seems that lsp-mode liberally expands the context collection range to the span at least the whole line after being asked to provide code actions on a point.

Makes sense. According to the spec, `context` is

```
	 * An array of diagnostics known on the client side overlapping the range
	 * provided to the `textDocument/codeAction` request.
```

Since the range is the whole line, it seems reasonable (and maybe even expected?) to send all diagnostics in the line.

> As far as I can read from the code, yes Eglot sends context, of course it does.

Oh, sorry about the confusion. I didn't look at the code and assumed it never send it just based on that single request.

**Reply by joaotavora** [reply]
> No, eglot is sending empty diagnostics as context – this is present in the logs I put in the discussion.

Yes, but I asked if the diagnostic exists at all.  It seems that it does, Eglot is just not sending it as part of its code actions request.

> Since the range is the whole line, it seems reasonable (and maybe even expected?) to send all diagnostics in the line.

But the thing is that `lsp-mode` is sending the same 1-character long `range` 
```
"range": { "start": { "line": 2, "character": 12 }, "end": { "line": 2,  "character": 12 }}
```
in the request. This doesn't overlap _any_ diagnostics, as far as I can tell.

`lsp-mode` seems to "liberally" (read "outside of the spec you posted") sends diagnostics that are _outside_ that 1-character long range, like the one between characters 1 and 8.  And this is a liberty that Eglot doesn't take, because it seems to be squarely outside the spec.

So if you were to invoke `M-x eglot-code-actions` with point somewhere between columns 1 and 8 of line 2, instead of requesting them in column 12, then I think even now, Eglot would provide the whole rhange.

What we _could_ do is make it so that, by default, if no region is selected, Eglot takes the whole line to be the range when `eglot-code-actions` is called interactively via `M-x` or via some keybinding.  Then it would be inside the spec and still function "intuitively".  

That is _if_ we agree that `lsp-mode` is more intuitive...  I'm not very sure it is. Why don't you invoke code actions between character 1 and 8?  Why do you expect that if point is at the end of 

```clojure
(str/foo) bar
```

you should get code actions for fixing `(str/foo)`?  I mean, it happens to be on the same line, but it _is_ a different lisp form/symbolic expression.

**Reply by Andre0991** [reply]
> Yes, but I asked if the diagnostic exists at all. It seems that it does, Eglot is just not sending it as part of its code actions request.

Got it. That's correct, it does exist.

> But the thing is that lsp-mode is sending the same 1-character long range

Oh, you're totally right. I though it started with 0 🤦 Sorry.

On the motivation for this change:

It's convenient to get code actions for a diagnostic without having the point above it. Usually, I execute `consult-flymake` for getting to the diagnostics. With this change, if the point is already in the same line, presenting code actions for the whole range would save a few keystrokes. Not a big deal, but a bit easier.

More importantly, importing namespaces would be much smoother. In particular, when you just wrote something with a diagnostic that offers an import code action, you don't have to manually backward one char for getting the code action. I hope this gif explain this situation:

![Mar-24-2022 09-55-24](https://user-images.githubusercontent.com/7083061/159921023-0225d98b-f4de-40fd-ae59-a6594ca6f003.gif)

Note that I write `set/join`, ask for code action and the 'require namespace' is not there.
Then, I move the point back and I'm able to get the expected code action.

**Reply by joaotavora** [reply]
I understand . The range for the whole line I think is too large, and too arbitrary  (especially for lisps, right? ;-) ) But the range for the current "thing at point" makes sense to me.  Can you tell me what this 100% untested patch to `eglot.el` yields for you?

```diff
diff --git a/eglot.el b/eglot.el
index ea9299a..1d3070d 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2821,8 +2821,10 @@ is not active."
                                            :newName ,newname))
    current-prefix-arg))
 
-(defun eglot--region-bounds () "Region bounds if active, else point and nil."
-  (if (use-region-p) `(,(region-beginning) ,(region-end)) `(,(point) nil)))
+(defun eglot--region-bounds ()
+  "Region bounds if active, else bounds of things at point."
+  (if (use-region-p) `(,(region-beginning) ,(region-end))
+    (bounds-of-thing-at-point 'sexp)))
 
 (defun eglot-code-actions (beg &optional end action-kind)
   "Offer to execute actions of ACTION-KIND between BEG and END.
```

**Reply by Andre0991** [reply]
> The range for the whole line I think is too large, and too arbitrary (especially for lisps, right? ;-) ) But the range for the current "thing at point" makes sense to me.

Fair point.

With the patch, when asking for code actions with the point just after the symbol (as in `(set/union|)`, where `|` is the point), I get this error:

```
byte-code: Wrong type argument: listp, 25462
```

Using debug on error:

```
Debugger entered--Lisp error: (wrong-type-argument listp 25462)
  append((25455 . 25462) (nil))
  byte-code("\301\302 \10\205\13\0\303\304\305\"C\"\207" [current-prefix-arg append eglot--region-bounds completing-read "[eglot] Action kind: " ("quickfix" "refactor.extract" "refactor.inline" "refactor.rewrite" "source.organizeImports")] 5)
  command-execute(eglot-code-actions)
```

**Reply by joaotavora** [reply]
Sorry about that, the usual Lisp brainfart here. Try this one:
```diff
diff --git a/eglot.el b/eglot.el
index ea9299a..29722b2 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2821,8 +2821,11 @@ is not active."
                                            :newName ,newname))
    current-prefix-arg))
 
-(defun eglot--region-bounds () "Region bounds if active, else point and nil."
-  (if (use-region-p) `(,(region-beginning) ,(region-end)) `(,(point) nil)))
+(defun eglot--region-bounds ()
+  "Region bounds if active, else bounds of things at point."
+  (if (use-region-p) `(,(region-beginning) ,(region-end))
+    (let ((botap (bounds-of-thing-at-point 'sexp)))
+      (list (car botap) (cdr botap)))))
 
 (defun eglot-code-actions (beg &optional end action-kind)
   "Offer to execute actions of ACTION-KIND between BEG and END.
```

**Reply by Andre0991** [reply]
This works as intended, yes. I tested with and without region 💯

**Reply by joaotavora** [reply]
OK I pushed this to master.  I think it makes sense.  Let's see if there aren't many complaints :smile:

**Reply by Andre0991** [reply]
Perfect. Many thanks!

**Comment by Andre0991** [comment]
Resolved by https://github.com/joaotavora/eglot/commit/03fc783c4b701fc8c19096b7167b73bd5d8f63a8.
`eglot` considers the 'thing at point' now.

Thanks, João.
